elasticsearch/src/main/scala/akka/stream/alpakka/elasticsearch/impl/ElasticsearchFlowStage.scala

```diff
@@ -44,137 +48,77 @@ private[elasticsearch] final class ElasticsearchFlowStage[T, C](
-      // Retry partial failed messages
-      // NOTE: When we partially return message like this, message will arrive out of order downstream
-      // and it can break commit-logic when using Kafka
-
-      emitResults(successMsgsIt.map(_._1).toList)
-      failedMsgs.groupBy(_._2).foreach {
-        case (retryCount, failedMsgsByRetry) =>
-          if (settings.retryLogic.shouldRetry(retryCount, Nil)) {
-            val updatedFailedMsgsByRetry = failedMsgsByRetry.map {
-              case (wr, _) =>
-                WriteMessageWithRetry(wr.message, retryCount + 1)
-            }
-            scheduleOnce(RetrySend(updatedFailedMsgsByRetry), settings.retryLogic.nextRetry(retryCount))
-          } else {
-            emitResults(failedMsgsByRetry.map(_._1))
-          }
-      }
-    }
+    setHandlers(in, out, this)
 
-    private def emitResults(successMsgs: immutable.Seq[WriteResult[T, C]]): Unit = {
-      emit(out, successMsgs)
-      tryPull()
-      inflight -= successMsgs.size
-      if (upstreamFinished && inflight == 0) completeStage()
-    }
+    override def onPull(): Unit = tryPull()
 
-    private def sendBulkUpdateRequest(messages: immutable.Seq[WriteMessageWithRetry[T, C]]): Unit = {
-      val json: String = restApi.toJson(messages.map(_.writeMessage))
+    override def onPush(): Unit = {
+      val (messages, resultsPassthrough) = grab(in)
+      inflight = true
+      val json: String = restApi.toJson(messages)
 
       log.debug("Posting data to Elasticsearch: {}", json)
 
+      // https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/java-rest-low-usage-requests.html
       client.performRequestAsync(
         "POST",
         "/_bulk",
         java.util.Collections.emptyMap[String, String](),
         new StringEntity(json, StandardCharsets.UTF_8),
         new ResponseListener() {
-          override def onFailure(exception: Exception): Unit = failureHandler.invoke((messages, exception))
-          override def onSuccess(response: Response): Unit = responseHandler.invoke((messages, response))
+          override def onFailure(exception: Exception): Unit =
+            failureHandler.invoke((resultsPassthrough, exception))
+
+          override def onSuccess(response: Response): Unit =
+            responseHandler.invoke((messages, resultsPassthrough, response))
         },
         new BasicHeader("Content-Type", "application/x-ndjson")
       )
     }
 
-    setHandlers(in, out, this)
+    private def handleFailure(
+        args: (immutable.Seq[WriteResult[T, C]], Throwable)
+    ): Unit = {
+      inflight = false
+      val (resultsPassthrough, exception) = args
 
-    override def onPull(): Unit = tryPull()
+      log.error(s"Received error from elastic after having already processed {} documents. Error: {}",
+                resultsPassthrough.size,
+                exception)
+      failStage(exception)
```

Comment:
Earlier it would retry even for errors, did that ever make sense?

Reply:
I initially considered it, however:
1. In order to encode the exception in a `WriteResult` I would need to abuse the `error` passing the exception message. Downstream stages would have no (clean) way to discriminate between valid Elasticsearch errors and exceptions.
2. The old code would fail the stage if the error persisted throughout all the retries. However having delegated the retry logic to the `RetryFlow` I have no way to make that decision after all the retries have been exhausted. And due to 1 I cannot easily check for this kind of failure in a downstream flow.
 
I think that its better to let the stage fail for this kind of errors happens and maybe wrap the flow in a `RestartFlow.onFailuresWithBackoff`.

Personally this is something I prefer to handle in my client code instead of having the library do it for me, but if you think its better to have this handled transparently we can do something like:

```
  @InternalApi
  private def basicStageFlow[T, C](indexName: String,
                                   typeName: String,
                                   settings: ElasticsearchWriteSettings,
                                   elasticsearchClient: RestClient,
                                   writer: MessageWriter[T]) = {
    def flowFactory() = Flow.fromGraph {
      new impl.ElasticsearchSimpleFlowStage[T, C](indexName, typeName, elasticsearchClient, settings, writer)
    }
    
    RestartFlow.withBackoff(settings.retryLogic.minBackoff, settings.retryLogic.maxBackoff, 0, settings.retryLogic.maxRetries)(flowFactory)
  }
```

Reply:
Fair enough. I agree that is is better to let the user decide about that retrying.